File: native-ios/Solvio/Core/UI/ScanQueueWidget.swift
import SwiftUI

/// Floating chip pinned above the bottom tab bar that surfaces the
/// `ScanQueueManager` state. Tap to expand into a panel listing every
/// queued / processing / saved / failed receipt.
///
/// Visual language matches the rest of the neobrutalism UI: hard borders,
/// drop shadow, mono labels, no spinners that block the user from working
/// with the rest of the app.
struct ScanQueueWidget: View {
    @EnvironmentObject private var queue: ScanQueueManager
    @EnvironmentObject private var locale: AppLocale
    @EnvironmentObject private var router: AppRouter
    @State private var expanded = false

    var body: some View {
        if queue.hasActivity {
            VStack(alignment: .leading, spacing: 0) {
                if expanded {
                    expandedPanel
                        .transition(.move(edge: .bottom).combined(with: .opacity))
                }
                chip
            }
            .padding(.horizontal, Theme.Spacing.sm)
            .animation(.spring(response: 0.32, dampingFraction: 0.85), value: expanded)
            .animation(.spring(response: 0.32, dampingFraction: 0.85), value: queue.items.count)
        }
    }

    // MARK: - Chip

    private var chip: some View {
        Button {
            expanded.toggle()
        } label: {
            HStack(spacing: Theme.Spacing.sm) {
                statusBadge
                VStack(alignment: .leading, spacing: 2) {
                    Text(headlineText)
                        .font(AppFont.bodyMedium)
                        .foregroundColor(Theme.foreground)
                        .lineLimit(1)
                    Text(subheadText)
                        .font(AppFont.mono(10))
                        .tracking(0.5)
                        .textCase(.uppercase)
                        .foregroundColor(Theme.mutedForeground)
                        .lineLimit(1)
                }
                Spacer(minLength: 0)
                if queue.hasInFlight {
                    progressRing
                } else if queue.failedCount > 0 {
                    Image(systemName: "exclamationmark.triangle.fill")
                        .font(.system(size: 16, weight: .bold))
                        .foregroundColor(Theme.destructive)
                } else {
                    Image(systemName: "checkmark.circle.fill")
                        .font(.system(size: 18, weight: .bold))
                        .foregroundColor(Theme.success)
                }
                Image(systemName: expanded ? "chevron.down" : "chevron.up")
                    .font(.system(size: 12, weight: .bold))
                    .foregroundColor(Theme.mutedForeground)
            }
            .padding(.horizontal, Theme.Spacing.sm)
            .padding(.vertical, Theme.Spacing.xs)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Theme.card)
            .clipShape(RoundedRectangle(cornerRadius: Theme.Radius.md))
            .overlay(
                RoundedRectangle(cornerRadius: Theme.Radius.md)
                    .stroke(Theme.border, lineWidth: Theme.Border.widthThin)
            )
            .nbShadow(Theme.Shadow.md)
        }
        .buttonStyle(.plain)
    }

    private var statusBadge: some View {
        ZStack {
            RoundedRectangle(cornerRadius: Theme.Radius.sm)
                .fill(Theme.foreground)
                .frame(width: 36, height: 36)
            Image(systemName: queue.hasInFlight ? "doc.text.viewfinder" : "tray.full.fill")
                .font(.system(size: 16, weight: .bold))
                .foregroundColor(Theme.background)
        }
    }

    private var progressRing: some View {
        ZStack {
            Circle()
                .stroke(Theme.border.opacity(0.2), lineWidth: 3)
                .frame(width: 22, height: 22)
            Circle()
                .trim(from: 0, to: max(0.05, queue.progress))
                .stroke(Theme.border, style: StrokeStyle(lineWidth: 3, lineCap: .round))
                .rotationEffect(.degrees(-90))
                .frame(width: 22, height: 22)
                .animation(.easeOut(duration: 0.25), value: queue.progress)
        }
    }

    private var headlineText: String {
        if queue.hasInFlight {
            let total = queue.items.count
            let done = queue.items.filter(\.status.isTerminal).count
            return "\(locale.t("scanQueue.scanning")) \(done)/\(total)"
        }
        if queue.failedCount > 0 {
            return locale.t("scanQueue.someFailed")
        }
        return locale.t("scanQueue.allDone")
    }

    private var subheadText: String {
        if queue.hasInFlight {
            return locale.t("scanQueue.tapToView")
        }
        return "\(queue.savedCount) \(locale.t("scanQueue.saved"))"
            + (queue.failedCount > 0 ? " · \(queue.failedCount) \(locale.t("scanQueue.failed"))" : "")
    }

    // MARK: - Expanded panel

    private var expandedPanel: some View {
        VStack(alignment: .leading, spacing: Theme.Spacing.xs) {
            HStack {
                Text(locale.t("scanQueue.title"))
                    .font(AppFont.mono(11))
                    .tracking(1)
                    .textCase(.uppercase)
                    .foregroundColor(Theme.mutedForeground)
                Spacer()
                if queue.items.contains(where: \.status.isTerminal) {
                    Button {
                        queue.clearCompleted()
                    } label: {
                        Text(locale.t("scanQueue.clearDone"))
                            .font(AppFont.mono(10))
                            .tracking(0.5)
                            .textCase(.uppercase)
                            .foregroundColor(Theme.foreground)
                            .padding(.horizontal, 8)
                            .padding(.vertical, 4)
                            .background(Theme.muted)
                            .clipShape(RoundedRectangle(cornerRadius: 4))
                            .overlay(
                                RoundedRectangle(cornerRadius: 4)
                                    .stroke(Theme.border.opacity(0.4), lineWidth: Theme.Border.widthThin)
                            )
                    }
                    .buttonStyle(.plain)
                }
            }

            ScrollView(showsIndicators: false) {
                VStack(spacing: 6) {
                    ForEach(queue.items.reversed()) { item in
                        row(for: item)
                    }
                }
            }
            .frame(maxHeight: 240)
        }
        .padding(Theme.Spacing.sm)
        .background(Theme.background)
        .clipShape(RoundedRectangle(cornerRadius: Theme.Radius.md))
        .overlay(
            RoundedRectangle(cornerRadius: Theme.Radius.md)
                .stroke(Theme.border, lineWidth: Theme.Border.widthThin)
        )
        .nbShadow(Theme.Shadow.md)
        .padding(.bottom, 6)
    }

    @ViewBuilder
    private func row(for item: ScanQueueManager.ScanQueueItem) -> some View {
        HStack(spacing: Theme.Spacing.sm) {
            Image(uiImage: item.thumbnail)
                .resizable()
                .scaledToFill()
                .frame(width: 36, height: 36)
                .clipShape(RoundedRectangle(cornerRadius: Theme.Radius.sm))
                .overlay(
                    RoundedRectangle(cornerRadius: Theme.Radius.sm)
                        .stroke(Theme.border.opacity(0.4), lineWidth: Theme.Border.widthThin)
                )

            VStack(alignment: .leading, spacing: 2) {
                Text(rowTitle(for: item))
                    .font(AppFont.bodyMedium)
                    .foregroundColor(Theme.foreground)
                    .lineLimit(1)
                Text(rowSubtitle(for: item))
                    .font(AppFont.caption)
                    .foregroundColor(Theme.mutedForeground)
                    .lineLimit(1)
            }

            Spacer(minLength: 0)

            statusIcon(for: item)
                .frame(width: 24, height: 24)

            if case .failed = item.status {
                Button {
                    queue.retry(id: item.id)
                } label: {
                    Image(systemName: "arrow.clockwise")
                        .font(.system(size: 12, weight: .bold))
                        .foregroundColor(Theme.foreground)
                        .frame(width: 28, height: 28)
                        .background(Theme.muted)
                        .clipShape(RoundedRectangle(cornerRadius: 6))
                        .overlay(
                            RoundedRectangle(cornerRadius: 6)
                                .stroke(Theme.border.opacity(0.4), lineWidth: Theme.Border.widthThin)
                        )
                }
                .buttonStyle(.plain)
            }
        }
        .padding(.horizontal, Theme.Spacing.xs)
        .padding(.vertical, 6)
        .background(Theme.muted.opacity(0.5))
        .clipShape(RoundedRectangle(cornerRadius: Theme.Radius.sm))
        .contentShape(Rectangle())
        .onTapGesture {
            if case .saved = item.status, let rid = item.receiptId {
                expanded = false
                router.push(.receiptDetail(id: rid))
            }
        }
    }

    private func rowTitle(for item: ScanQueueManager.ScanQueueItem) -> String {
        if let v = item.vendor, !v.isEmpty { return v }
        switch item.status {
        case .pending: return locale.t("scanQueue.queued")
        case .uploading: return locale.t("scanQueue.uploading")
        case .processing: return locale.t("scanQueue.processing")
        case .saved: return locale.t("receipts.savedScan")
        case .failed: return locale.t("scanQueue.failedShort")
        }
    }

    private func rowSubtitle(for item: ScanQueueManager.ScanQueueItem) -> String {
        if let total = item.total, let cur = item.currency {
            return Fmt.amount(total, currency: cur)
        }
        switch item.status {
        case .pending: return locale.t("scanQueue.waitingTurn")
        case .uploading: return locale.t("scanQueue.uploadingDetail")
        case .processing: return locale.t("scanQueue.processingDetail")
        case .saved: return locale.t("scanQueue.tapToOpen")
        case .failed(let m): return m
        }
    }

    @ViewBuilder
    private func statusIcon(for item: ScanQueueManager.ScanQueueItem) -> some View {
        switch item.status {
        case .pending:
            Image(systemName: "clock")
                .font(.system(size: 14, weight: .bold))
                .foregroundColor(Theme.mutedForeground)
        case .uploading, .processing:
            ProgressView()
                .progressViewStyle(.circular)
                .scaleEffect(0.7)
                .tint(Theme.foreground)
        case .saved:
            Image(systemName: "checkmark.circle.fill")
                .font(.system(size: 16, weight: .bold))
                .foregroundColor(Theme.success)
        case .failed:
            Image(systemName: "xmark.circle.fill")
                .font(.system(size: 16, weight: .bold))
                .foregroundColor(Theme.destructive)
        }
    }
}

File: android/app/src/main/java/com/programo/solvio/core/ui/ScanQueueWidget.kt
package com.programo.solvio.core.ui

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.expandVertically
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.shrinkVertically
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Description
import androidx.compose.material.icons.filled.ErrorOutline
import androidx.compose.material.icons.filled.ExpandLess
import androidx.compose.material.icons.filled.ExpandMore
import androidx.compose.material.icons.filled.Inventory2
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.Schedule
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.unit.dp
import com.programo.solvio.LocalAppLocale
import com.programo.solvio.core.Fmt
import com.programo.solvio.core.ScanQueueManager
import com.programo.solvio.core.theme.LocalPalette
import com.programo.solvio.core.theme.SolvioFonts
import com.programo.solvio.core.theme.SolvioTheme
import com.programo.solvio.core.theme.nbShadow

/// Floating chip pinned above the bottom tab bar that surfaces the
/// ScanQueueManager state. Tap to expand into a panel listing every
/// queued / processing / saved / failed receipt.
///
/// Visual language matches the rest of the neobrutalism UI: hard borders,
/// drop shadow, mono labels, no spinners that block the user.
@Composable
fun ScanQueueWidget(
    queue: ScanQueueManager = ScanQueueManager.get(),
    onOpenReceipt: (String) -> Unit = {},
    modifier: Modifier = Modifier,
) {
    val palette = LocalPalette.current
    val locale = LocalAppLocale.current
    val items by queue.items.collectAsState()
    var expanded by remember { mutableStateOf(false) }

    if (items.isEmpty()) return

    val hasInFlight = items.any { !it.status.isTerminal }
    val saved = items.count { it.status is ScanQueueManager.Status.Saved }
    val failed = items.count { it.status is ScanQueueManager.Status.Failed }
    val done = items.count { it.status.isTerminal }

    Column(modifier = modifier.padding(horizontal = SolvioTheme.Spacing.sm)) {
        AnimatedVisibility(
            visible = expanded,
            enter = fadeIn() + expandVertically(),
            exit = fadeOut() + shrinkVertically(),
        ) {
            ExpandedPanel(
                items = items,
                onClear = { queue.clearCompleted() },
                onRetry = { id -> queue.retry(id) },
                onTapItem = { id -> onOpenReceipt(id); expanded = false },
            )
        }

        // Chip
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .nbShadow(palette, offset = SolvioTheme.Shadow.md)
                .clip(RoundedCornerShape(SolvioTheme.Radius.md))
                .background(palette.surface)
                .border(SolvioTheme.Border.widthThin, palette.border, RoundedCornerShape(SolvioTheme.Radius.md))
                .clickable { expanded = !expanded }
                .padding(horizontal = SolvioTheme.Spacing.sm, vertical = SolvioTheme.Spacing.xs),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Box(
                modifier = Modifier
                    .size(36.dp)
                    .clip(RoundedCornerShape(SolvioTheme.Radius.sm))
                    .background(palette.foreground),
                contentAlignment = Alignment.Center,
            ) {
                Icon(
                    imageVector = if (hasInFlight) Icons.Filled.Description else Icons.Filled.Inventory2,
                    contentDescription = null,
                    tint = palette.background,
                    modifier = Modifier.size(18.dp),
                )
            }
            Spacer(Modifier.width(SolvioTheme.Spacing.sm))
            Column(modifier = Modifier.weight(1f)) {
                val headline = when {
                    hasInFlight -> "${locale.t("scanQueue.scanning")} $done/${items.size}"
                    failed > 0 -> locale.t("scanQueue.someFailed")
                    else -> locale.t("scanQueue.allDone")
                }
                val sub = if (hasInFlight) {
                    locale.t("scanQueue.tapToView")
                } else {
                    "$saved ${locale.t("scanQueue.saved")}" +
                        if (failed > 0) " · $failed ${locale.t("scanQueue.failed")}" else ""
                }
                Text(headline, style = SolvioFonts.bodyMedium.copy(color = palette.foreground), maxLines = 1)
                Text(sub.uppercase(), style = SolvioFonts.mono(10).copy(color = palette.mutedForeground), maxLines = 1)
            }
            when {
                hasInFlight -> CircularProgressIndicator(
                    color = palette.foreground,
                    strokeWidth = 2.dp,
                    modifier = Modifier.size(20.dp),
                )
                failed > 0 -> Icon(Icons.Filled.ErrorOutline, contentDescription = null, tint = palette.destructive)
                else -> Icon(Icons.Filled.CheckCircle, contentDescription = null, tint = palette.success)
            }
            Spacer(Modifier.width(8.dp))
            Icon(
                imageVector = if (expanded) Icons.Filled.ExpandMore else Icons.Filled.ExpandLess,
                contentDescription = null,
                tint = palette.mutedForeground,
                modifier = Modifier.size(16.dp),
            )
        }
    }
}

@Composable
private fun ExpandedPanel(
    items: List<ScanQueueManager.Item>,
    onClear: () -> Unit,
    onRetry: (String) -> Unit,
    onTapItem: (String) -> Unit,
) {
    val palette = LocalPalette.current
    val locale = LocalAppLocale.current

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(bottom = 6.dp)
            .nbShadow(palette, offset = SolvioTheme.Shadow.md)
            .clip(RoundedCornerShape(SolvioTheme.Radius.md))
            .background(palette.background)
            .border(SolvioTheme.Border.widthThin, palette.border, RoundedCornerShape(SolvioTheme.Radius.md))
            .padding(SolvioTheme.Spacing.sm),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(
                locale.t("scanQueue.title").uppercase(),
                style = SolvioFonts.mono(11).copy(color = palette.mutedForeground),
                modifier = Modifier.weight(1f),
            )
            if (items.any { it.status.isTerminal }) {
                Box(
                    modifier = Modifier
                        .clip(RoundedCornerShape(4.dp))
                        .background(palette.muted)
                        .border(SolvioTheme.Border.widthThin, palette.border.copy(alpha = 0.4f), RoundedCornerShape(4.dp))
                        .clickable { onClear() }
                        .padding(horizontal = 8.dp, vertical = 4.dp),
                ) {
                    Text(
                        locale.t("scanQueue.clearDone").uppercase(),
                        style = SolvioFonts.mono(10).copy(color = palette.foreground),
                    )
                }
            }
        }
        Spacer(Modifier.height(SolvioTheme.Spacing.xs))
        val ordered = items.asReversed()
        LazyColumn(
            verticalArrangement = Arrangement.spacedBy(6.dp),
            modifier = Modifier.heightIn(max = 240.dp),
        ) {
            items(items = ordered, key = { it.id }) { item ->
                QueueRow(item = item, onRetry = onRetry, onTap = onTapItem)
            }
        }
    }
}

@Composable
private fun QueueRow(
    item: ScanQueueManager.Item,
    onRetry: (String) -> Unit,
    onTap: (String) -> Unit,
) {
    val palette = LocalPalette.current
    val locale = LocalAppLocale.current

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(SolvioTheme.Radius.sm))
            .background(palette.muted.copy(alpha = 0.5f))
            .clickable(enabled = item.status is ScanQueueManager.Status.Saved && item.receiptId != null) {
                item.receiptId?.let { onTap(it) }
            }
            .padding(horizontal = SolvioTheme.Spacing.xs, vertical = 6.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Image(
            bitmap = item.thumbnail.asImageBitmap(),
            contentDescription = null,
            modifier = Modifier
                .size(36.dp)
                .clip(RoundedCornerShape(SolvioTheme.Radius.sm))
                .border(SolvioTheme.Border.widthThin, palette.border.copy(alpha = 0.4f), RoundedCornerShape(SolvioTheme.Radius.sm)),
        )
        Spacer(Modifier.width(SolvioTheme.Spacing.sm))
        Column(modifier = Modifier.weight(1f)) {
            val title = item.vendor?.takeIf { it.isNotBlank() } ?: when (item.status) {
                is ScanQueueManager.Status.Pending -> locale.t("scanQueue.queued")
                is ScanQueueManager.Status.Uploading -> locale.t("scanQueue.uploading")
                is ScanQueueManager.Status.Processing -> locale.t("scanQueue.processing")
                is ScanQueueManager.Status.Saved -> locale.t("receipts.savedScan")
                is ScanQueueManager.Status.Failed -> locale.t("scanQueue.failedShort")
            }
            val subtitle: String = if (item.total != null && item.currency != null) {
                Fmt.amount(item.total, item.currency)
            } else when (val s = item.status) {
                is ScanQueueManager.Status.Pending -> locale.t("scanQueue.waitingTurn")
                is ScanQueueManager.Status.Uploading -> locale.t("scanQueue.uploadingDetail")
                is ScanQueueManager.Status.Processing -> locale.t("scanQueue.processingDetail")
                is ScanQueueManager.Status.Saved -> locale.t("scanQueue.tapToOpen")
                is ScanQueueManager.Status.Failed -> s.message
            }
            Text(title, style = SolvioFonts.bodyMedium.copy(color = palette.foreground), maxLines = 1)
            Text(subtitle, style = SolvioFonts.caption.copy(color = palette.mutedForeground), maxLines = 1)
        }
        Spacer(Modifier.width(SolvioTheme.Spacing.xs))
        // Status icon
        Box(modifier = Modifier.size(24.dp), contentAlignment = Alignment.Center) {
            when (item.status) {
                is ScanQueueManager.Status.Pending -> Icon(Icons.Filled.Schedule, contentDescription = null, tint = palette.mutedForeground, modifier = Modifier.size(14.dp))
                is ScanQueueManager.Status.Uploading,
                is ScanQueueManager.Status.Processing -> CircularProgressIndicator(
                    color = palette.foreground,
                    strokeWidth = 2.dp,
                    modifier = Modifier.size(16.dp),
                )
                is ScanQueueManager.Status.Saved -> Icon(Icons.Filled.CheckCircle, contentDescription = null, tint = palette.success, modifier = Modifier.size(16.dp))
                is ScanQueueManager.Status.Failed -> Icon(Icons.Filled.ErrorOutline, contentDescription = null, tint = palette.destructive, modifier = Modifier.size(16.dp))
            }
        }
        if (item.status is ScanQueueManager.Status.Failed) {
            Spacer(Modifier.width(6.dp))
            Box(
                modifier = Modifier
                    .size(28.dp)
                    .clip(RoundedCornerShape(6.dp))
                    .background(palette.muted)
                    .border(SolvioTheme.Border.widthThin, palette.border.copy(alpha = 0.4f), RoundedCornerShape(6.dp))
                    .clickable { onRetry(item.id) },
                contentAlignment = Alignment.Center,
            ) {
                Icon(Icons.Filled.Refresh, contentDescription = null, tint = palette.foreground, modifier = Modifier.size(14.dp))
            }
        }
    }
}
